### DISCUSSION TITLE
Using stackalloc inside of loops

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/runtime#63612] [Q&A] [answered]
If I use ```stackalloc``` inside of loops, I will receive a warning of [CA2014](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2014).
```
for (int i = 0; i < 10; i++)
{
    Span<int> test = stackalloc int[2]; //CA2014
    ...
}
```

About rule description of CA2014
> The C# stackalloc expression allocates memory from the current stack frame, and that memory may not be released until the current method call returns. If stackalloc is used in a loop, it can lead to stack overflows due to exhausting the stack memory.

I dont konw why the memory allocated by ```stackalloc``` is not released at the end of each cycle. Maybe, I should using ```stackalloc``` inside of local functions.
```
for (int i = 0; i < 10; i++)
{
    void SomeName()
    {
        Span<int> test = stackalloc int[2];
        ...
    }

    SomeName();
}
```

**Comment by Joe4evr** [comment]
The other option is to allocate your buffer outside of the loop and continuously reuse it. If stale data in the buffer is a concern, you can simply call `.Clear()`.

**Comment by YairHalberstadt** [comment, marked as answer]
I think it doesn't reuse the stackalloc so that this works:

```csharp
    var a = 0;
    var b = 0;
    ref int refA = ref a;
    ref int refB = ref b;

    foreach (var i in new[]{1,2})
    {
        Span<int> span = stackalloc int[]{i};
        if (i == 1)
        {
            refA = ref span[0];
        }
        else
        {
            refB = ref span[0];
        }
    }
    
    Console.WriteLine(refA);
    Console.WriteLine(refB);
```

This prints 1,2 instead of 2,2 as it would if the stackalloc were reused.

**Reply by yaakov-h** [reply]
This feels a bit like chicken-and-egg being the wrong way around to me. `stackalloc`'s behaviour predates `Span` - every time we `stackalloc`, we allocate new memory on the stack, but there's no `stackfree` - rather it is up to normal stack operations to release the memory, i.e. the end of the method.

The fact that we keep taking the pointer to memory allocated by `stackalloc` and wrapping it with a `Span` and storing it in the same variable is irrelevant to the runtime. It got asked to allocate more memory, so it did.

**Reply by YairHalberstadt** [reply]
Fair, I guess I was giving an example where changing this would be a breaking change.

**Reply by Joe4evr** [reply]
You could write [the example in terms of pointers](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4AEBMBGAsAKBQZgAJ1CBhXAb10JuKIFcA7AZwEM4JiAWQgWQAoAlNVpUctCYTCsohVoQC8hAAwBuEZKkzCAI0Ur14zYRiNgAKkIAHYFACC+gGStDxk2cs2oAISc7XxhqScAD2UBCsAMYAFoT80rIw7oSMEADuANoAuoQUhBgANIRohAC+wkaaYm4Spha69HAcskrMwFEA1qwANt0hke7A2bkmZQE1JnBxSQpKGBUTNNWLEl4OSo46jc0ZylnjNaVBbhDdzBDHxssrNF6+G1tN0Lv7l5pHlZIfbm/EGACc/HMa0EBwkKABQLuoKCH1KQA===), that should work the same even going back to long before `Span` and friends were introduced.

**Reply by gfoidl** [reply]
This it suffers from the same latent problems as outlined in https://github.com/dotnet/runtime/discussions/63612#discussioncomment-1945449. So don't do that.

**Reply by ghost** [reply]
> we allocate new memory on the stack, but there's no stackfree - rather it is up to normal stack operations to release the memory, i.e. the end of the method.

Thanks

**Reply by Joe4evr** [reply]
> This it suffers from the same latent problems as outlined in [#63612 (reply in thread)](https://github.com/dotnet/runtime/discussions/63612#discussioncomment-1945449). So don't do that.

I know, I only wrote it to showcase the problem is not exclusive to the use of `Span<T>` and `ref` returns/locals and such.

**Reply by tannergooding** [reply]
> I think it doesn't reuse the stackalloc so that this works:

This is an "implementation detail" and is technically relying on undefined behavior.

There is no strict requirement in ECMA 335 (or most language specs) that would require the stackalloc to be scoped to the entire method. As such, a sufficiently intelligent compiler would be free to alias the stackallocs, as it already can do (and does) for other types of locals.

**Comment by tannergooding** [comment]
> I dont konw why the memory allocated by stackalloc is not released at the end of each cycle.

Due to how stackalloc and stacks in general work, this is not always trivial. There are many compilers that simply let a stackalloc live until the method ends at which point it is "implicitly" freed as part of returning to the caller.
